File: forecasto-server/src/forecasto/schemas/prompt_builder.py
"""Prompt builder schemas."""

from __future__ import annotations

from datetime import datetime

from pydantic import BaseModel


class GeneratePromptRequest(BaseModel):
    """Request to generate/update a prompt."""

    force_regenerate: bool = False


class PromptUsageInfo(BaseModel):
    """Token usage details for a generation."""

    input_tokens: int
    output_tokens: int
    total_cost_eur: float
    model: str


class GeneratePromptResponse(BaseModel):
    """Response from prompt generation."""

    success: bool = True
    prompt: str
    is_update: bool = False
    usage: PromptUsageInfo
    records_analyzed: int


class AgentPromptResponse(BaseModel):
    """Response with current prompt status."""

    prompt: str | None = None
    last_generated_at: datetime | None = None
    records_analyzed: int = 0
    auto_update: bool = False
    records_since_regen: int = 0


class AgentPromptUpdate(BaseModel):
    """Manual prompt update."""

    prompt: str | None = None
    auto_update: bool | None = None


class PatternAnalysisResponse(BaseModel):
    """Raw pattern analysis results (no LLM cost)."""

    total_records: int
    account_frequency: list[dict]
    reference_account_mapping: list[dict]
    reference_total_patterns: list[dict]
    type_area_mapping: list[dict]
    vat_deduction_patterns: list[dict]
    withholding_patterns: list[dict]
    project_account_mapping: list[dict]
    stage_patterns: list[dict]
    payment_terms: list[dict]


class PromptGenerationJobResponse(BaseModel):
    """Single generation job for history."""

    id: str
    scope: str
    status: str
    llm_model: str | None
    input_tokens: int
    output_tokens: int
    total_cost_eur: float
    records_analyzed: int
    billing_month: str
    created_at: datetime

    model_config = {"from_attributes": True}


class UsageSummaryResponse(BaseModel):
    """Aggregated usage per month."""

    month: str
    total_input_tokens: int
    total_output_tokens: int
    total_cost_eur: float
    generation_count: int


class UsageSummaryListResponse(BaseModel):
    """List of monthly usage summaries."""

    months: list[UsageSummaryResponse]

File: forecasto-client-web/src/types/promptBuilder.ts
export interface PromptUsageInfo {
  input_tokens: number
  output_tokens: number
  total_cost_eur: number
  model: string
}

export interface GeneratePromptResponse {
  success: boolean
  prompt: string
  is_update: boolean
  usage: PromptUsageInfo
  records_analyzed: number
}

export interface AgentPromptResponse {
  prompt: string | null
  last_generated_at: string | null
  records_analyzed: number
}

export interface PatternAnalysisResponse {
  total_records: number
  account_frequency: Array<{ account: string; total: number; in_count: number; out_count: number }>
  reference_account_mapping: Array<{ reference: string; account: string; count: number }>
  reference_total_patterns: Array<{ reference: string; count: number; avg_total: number; min_total: number; max_total: number }>
  type_area_mapping: Array<{ type: string; area: string; count: number }>
  vat_deduction_patterns: Array<{ account: string; vat_deduction: number; count: number }>
  withholding_patterns: Array<{ type: string; account: string; withholding_rate: number; count: number }>
  project_account_mapping: Array<{ project_code: string; account: string; count: number }>
  stage_patterns: Array<{ reference: string; stage: string; count: number }>
  payment_terms: Array<{ reference: string; count: number; avg_days: number }>
}

export interface PromptGenerationJob {
  id: string
  scope: string
  status: string
  llm_model: string | null
  input_tokens: number
  output_tokens: number
  total_cost_eur: number
  records_analyzed: number
  billing_month: string
  created_at: string
}

export interface UsageSummary {
  month: string
  total_input_tokens: number
  total_output_tokens: number
  total_cost_eur: number
  generation_count: number
}
